datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -515,14 +522,15 @@ fn build_common_expr_project_plan(
     let mut fields_set = BTreeSet::new();
     let mut project_exprs = common_exprs
         .into_iter()
-        .map(|(expr_id, expr)| {
+        .enumerate()
+        .map(|(index, (expr_id, expr))| {
```

Comment:
Generating unique aliases with some kind of index is a great idea, but I'm not sure this implementation is robust enough. The rule `CommonSubexprEliminate` can run many times during optimization so starting alias index always from 0 might result colliding column names.

I wonder if we could check the schema for possible duplicates before assigning an alias? This idea has actually come up before: https://github.com/apache/datafusion/pull/10473#discussion_r1604645685, but would be better to implement in a separate PR like this one.

Also, do you think we can assign aliases with "cse_" or "common_" prefix? IMO we don't need a special `#` prefix for that.

Reply:
I can't give you a proof of why no collisions will happen as I'm yet trying to fully understand how this implementation works with all of the different optimization functions.

But FWIW I followed every reference of `expr_id`, `expr_identifier()` and `common_exprs` closely and made sure that this change will work for all the cases where the previous identifier did. 

The only problem I faced was with `try_optimize_aggregate` where the aggregations are handled by different functions, keeping track of the [last index](https://github.com/apache/datafusion/pull/10868/files#diff-351499880963d6a383c92e156e75019cd9ce33107724a9635853d7d4cd1898d0R332) seems to have solved the issue though.

I see no point in indicating to the user which rule generated that alias. If the user is that interested, he can always run `EXPLAIN VERBOSE`.

If the `#` is a problem, we can drop it and just use the number or even brackets, but it's simply just a short placeholder alias that is used by a few other popular engines.

Reply:
So I was thinking of a following example:
```
Projection: a + b, a + b, a + 2, a + 1 + 1
```
When this query gets optimized it might be the case that `CommonSubexprEliminate` runs first and extracts `a + b` to `#0` so the plan becomes:
```
Projection: #0 as "a + b", #0 as "a + b", a + 2, a + 1 + 1
  Projection: a + b as #0, a
```
But then some other rule can do contant folding so the plan becomes:
```
Projection: #0 as "a + b", #0 as "a + b", a + 2, a + 2
  Projection: a + b as #0, a
```
And then in the 2nd optimizer cycle `CommonSubexprEliminate` runs again and wants to extract `a + 2`, but as far as I see your PR uses the same 0 based aliases again in the 2nd+ runs so the plan would become invalid:
```
Projection: #0, #0, #0, #0
  Projection: a + b as #0, a + 2 as #0, a
    Projection: a + b as #0, a
```

I'm not fully familiar with all the optimizer rules in DataFusion, so there might not be any "constant folding" one in: https://github.com/apache/datafusion/blob/main/datafusion/optimizer/src/optimizer.rs#L239-L272, but I hope you get the idea why the current change can cause issues.

Reply:
If `#` prefix used in other engines then I'm fine with it.

Reply:
We shouldn't worry about that scenario because we make sure that all expressions are properly aliased before writing them. The closest example I can find of a similar scenario is [this](https://github.com/apache/datafusion/pull/10868/files#diff-351499880963d6a383c92e156e75019cd9ce33107724a9635853d7d4cd1898d0R1114). You'll see that the CSE is aliased to its original name.

So for the example you gave it would be the "constant folding" responsibility to alias `#0` in the intermediate projection to something else.

I can try passing `common_exprs` between different optimization calls if that will make you feel safer about the change, but that may be challenging given DataFusion's language of choice (aka our beloved Rust).

Reply:
I've updated the example plans a bit in my previous comment to show the issue with `#0` alias chosen 2 times (in 2 different optimizer rule runs).

Reply:
> We shouldn't worry about that scenario because we make sure that all expressions are properly aliased before writing them. The closest example I can find of a similar scenario is this or maybe even this. You'll see that the CSE is aliased to its original name and the name can be given to other expressions.

The problem is with the extracted expressions' aliases in the newly added `Projection` nodes.

Reply:
Every single rewrite contains an alias (except for already aliased nodes as we've discussed before). Rewrites done on the same plan node share the same `common_exprs` indexmap (or at least the index, check `try_optimize_aggregate`) so your example will be `a + b as #0, #0 as #1"

Reply:
Hmm, I'm not sure I get your comment. This issue is not `try_optimize_aggregate` related and I don't see why `as #1` alias gets added. 

But thanks, an example would be great. You can mimic the constant folding step by manually modifying the plan.

Reply:
We were not on the same page.

When I wrote the previous comment, I thought that the uppermost `#1` would be considered for the elimination alongside the `a+2`, thus producing the intermediate projection `#1 as #1, a + 2 as #2`.

The only issue here is, ExprMask skips single columns and therefore `#1` column is never considered and creates this abomination of a plan ...

```
Projection: c1, c2, c3, c4\
  Projection: #1 AS c1, #1 AS c2, #1 AS c3, #1 AS c4\
    Projection: test.c + Int32(2) AS #1, test.c\
      Projection: test.a + test.b AS #1, test.c\
        TableScan: test
```

Reply:
Yeah, eliminating single columns is no use.
I think you need to choose an alias for the extracted expression in `f_down()` of `CommonSubexprRewriter` based on the schema (check that there is no collision) and store the alias in `common_exprs`.

Reply:
I pushed a change, take a look and let me know.

If it still doesn't solve the issue/it itself has issues, I'm afraid I'll have to close this PR as I don't think I'll be able to finish it anytime soon.